src/kakarot/instructions/system_operations.cairo

```diff
@@ -811,6 +832,305 @@ namespace SystemOperations {
+        memory: model.Memory*,
+        state: model.State*,
+    }(evm: model.EVM*) -> model.EVM* {
+        alloc_locals;
+        // Stack
+        let (popped) = Stack.pop_n(3);
+        let authority = uint256_to_uint160(popped[0]);
+        let offset = popped[1];
+        let length = popped[2];
+
+        // Gas
+        // Calling `get_account` subsequently will make the account warm for the next interaction
+        let is_warm = State.is_account_warm(authority);
+        tempvar access_gas_cost = is_warm * Gas.WARM_ACCESS + (1 - is_warm) *
+            Gas.COLD_ACCOUNT_ACCESS;
+
+        // Charge memory access gas - will revert if offset.high | length.high != 0
+        let memory_expansion = Gas.memory_expansion_cost_saturated(
+            memory.words_len, offset, length
+        );
+        let evm = EVM.charge_gas(evm, memory_expansion.cost + access_gas_cost);
+        if (evm.reverted != FALSE) {
+            return evm;
+        }
+
+        // OPERATION
+        tempvar memory = new model.Memory(
+            word_dict_start=memory.word_dict_start,
+            word_dict=memory.word_dict,
+            words_len=memory_expansion.new_words_len,
+        );
+
+        // authority not EOA
+        let authority_account = State.get_account(authority);
+        if (authority_account.code_len != 0) {
+            Stack.push_uint128(0);
+            tempvar authorized = new model.Option(is_some=0, value=0);
+            tempvar message = new model.Message(
+                bytecode=evm.message.bytecode,
+                bytecode_len=evm.message.bytecode_len,
+                valid_jumpdests_start=evm.message.valid_jumpdests_start,
+                valid_jumpdests=evm.message.valid_jumpdests,
+                calldata=evm.message.calldata,
+                calldata_len=evm.message.calldata_len,
+                value=evm.message.value,
+                caller=evm.message.caller,
+                parent=evm.message.parent,
+                address=evm.message.address,
+                code_address=evm.message.code_address,
+                read_only=evm.message.read_only,
+                is_create=evm.message.is_create,
+                authorized=authorized,
+                depth=evm.message.depth,
+                env=evm.message.env,
+            );
+            return new model.EVM(
+                message=message,
+                return_data_len=evm.return_data_len,
+                return_data=evm.return_data,
+                program_counter=evm.program_counter,
+                stopped=evm.stopped,
+                gas_left=evm.gas_left,
+                gas_refund=evm.gas_refund,
+                reverted=evm.reverted,
+            );
+        }
+
+        let (mem_slice) = alloc();
+        let mem_slice_len = length.low;
+        Memory.load_n(mem_slice_len, mem_slice, offset.low);
+
+        // In the event length < 97, pad the memory slice with zeros
+        let is_length_lt_97 = is_le(mem_slice_len, 96);
```

Comment:
can we create a `pad_end` method in `array.utils` on the model of https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Global_Objects/String/padEnd that does this

```
func pad_end(arr_len: felt, arr: felt*, size: felt) {
    // ...
}
```

we could do a `pad_start` as well, though, YAGNI

Reply:
I don't really see a benefic to have

```rust
func pad_end(arr_len:felt, arr:felt*,  size:felt){
    memset(arr+arr_len, 0, size);
    return();
}
```

compared to simply `memset(arr+arr_len, 0, size);`

Reply:
the `padEnd` in js (and other languages) pad if necessary, it's more like "ensure lenght is at least"

for exemple

```
> "1234".padEnd(5)
'1234 '
> "1234".padEnd(2)
'1234'
```

so the idea is to remove the if here. You could also use the `array.slice` utils but it'd create a useless copy